### PR TITLE
Add license script

### PR DESCRIPTION
**Pull request by facuMH**
This PR introduces the the sonic LGPL and a script that automates adding it to all relevant files.

This PR contains the 3 relevant files to the automation of this process. That is the license header in plain text, a go script that checks all files and adds the license if they need it, and a test for the scrip.

NOTE: the header is missing in the `_test.go` file. But it is added in https://github.com/0xsoniclabs/sonic/pull/334.

### CODE REVIEW

File: scripts/license/add_license_header.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+				// remove lines up to this point
+				content = []byte(strings.Join(lines[i+1:], "\n"))
+				break
+			}
+		}
+	}
+
+	// Add header
+	newContent := licenseHeader + "\n" + string(content)
+	return os.WriteFile(path, []byte(newContent), 0222)
```
Comment (HerbertJordan) [inline scripts/license/add_license_header.go:158]: Is this script implicitly changing the permission of all my files to public writeable? And no read permissions?

It would be good to preserve the same file permissions. Wdyt?

File: scripts/license/add_license_header.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+func checkDoubleHeader(path, prefix string) error {
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return fmt.Errorf("failed to read %s: %v", path, err)
+	}
+	// we check the first 20 lines because the license header should be the top 15 lines.
```
Comment (HerbertJordan) [inline scripts/license/add_license_header.go:166]: This comment line reads like it does not belong here ...

File: scripts/license/add_license_header_test.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func Test_Recognizes_GethLicense(t *testing.T) {
+	// make a temporary file
+	tmpDir := t.TempDir()
+	originalFile, err := os.Create(tmpDir + "/test_geth_license")
```
Comment (HerbertJordan) [inline scripts/license/add_license_header_test.go:13]:
```suggestion
	originalFile, err := os.Create(filepath.Join(tmpDir, "test_geth_license"))
```

File: scripts/license/add_license_header_test.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+	require.NoError(t, err)
+	defer func() { _ = originalFile.Close() }()
```
Comment (HerbertJordan) [inline scripts/license/add_license_header_test.go:15]:
```suggestion
	defer originalFile.Close()
```

File: scripts/license/add_license_header_test.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+	require.NoError(t, err)
+
+	// Check for double license headers
+	err = processFiles(tmpDir, ".go", "//", license, false, false)
+	require.NoError(t, err)
+
+	content, err := os.ReadFile(tmpFile.Name())
+	require.NoError(t, err)
+	firstEmptyLine := false
+	for _, line := range string(content) {
```
Comment (HerbertJordan) [inline scripts/license/add_license_header_test.go:191]: I think here you are iterating through characters, not lines ...

File: scripts/license/add_license_header_test.go (hunk 4 above)
Comment (facuMH) [reply scripts/license/add_license_header_test.go:15]: one of our linters is `errcheck` which checks that all functions that return an error assign it to something. So we need to assignt it event if we are going to ignore it. That said, most of these defers will go away when I apply your suggestion of using `[os.WriteFile](https://pkg.go.dev/os#WriteFile)`.

File: scripts/license/add_license_header.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+	license, err := getLicenseHeader()
+	if err != nil {
+		log.Fatalf("Failed to get license header: %v\n", err)
+	}
+
+	rootDir, err := os.Getwd()
+	if err != nil {
+		log.Fatalf("Failed to get working directory: %v\n", err)
+	}
+	rootDir = filepath.Join(rootDir, "../..")
```
[the remark this answers is not in the document]
Comment (facuMH) [reply scripts/license/add_license_header.go]: https://github.com/0xsoniclabs/sonic/pull/335/commits/09c32263c130ecf09f82d8460b98c521b7be2bca

File: scripts/license/add_license_header.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+		fmt.Printf("Processing files with extension %s using prefix '%s'\n", ext, prefix)
+		err := processFiles(rootDir, ext, prefix, string(license), *checkOnly, *checkDoubleHeader)
+		if err != nil {
+			log.Fatalf("Error processing files with extension %s: %v\n", ext, err)
+		}
+	}
+	os.Exit(result)
+}
+
+func processFiles(root, ext, prefix, license string, checkOnly, doubleHeader bool) error {
```
[the remark this answers is not in the document]
Comment (facuMH) [reply scripts/license/add_license_header.go]: https://github.com/0xsoniclabs/sonic/pull/335/commits/52939a71f6fa738a632b06932a51492c3daaed4d

File: scripts/license/add_license_header.go (hunk 1 above)
Comment (facuMH) [reply scripts/license/add_license_header.go]: https://github.com/0xsoniclabs/sonic/pull/335/commits/3c89a0aec3e33779aed3ff7afe2bea5b3fbb7a75

File: scripts/license/add_license_header_test.go (hunk 5 above)
Comment (facuMH) [reply scripts/license/add_license_header_test.go]: https://github.com/0xsoniclabs/sonic/pull/335/commits/ea236981378815b28b94fb3717a15fd8ace6fb49

File: scripts/license/add_license_header_test.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+	require.Error(t, checkDoubleHeader(tmpFileName, "//"))
+}
+
+func Test_OnlyOneEmptyLineAfterHeader(t *testing.T) {
+	// make a temporary folder
+	tmpDir := t.TempDir()
+	// make a temporary file in that folder
+	tmpFileName := filepath.Join(tmpDir, "test_license.go")
+
+	fileWithHeader := addPrefix(licenseHeader, "//") + "\npackage main\nfunc main() {}\n"
```
Comment (HerbertJordan) [inline scripts/license/add_license_header_test.go:130]: It looks like this test is preparing a file with a proper license header. So the `processFiles` should not have any affect.

A more interesting test case would be if you start with a file with a few empty lines on top, an empty file, or a file with a doc-string on top.

File: scripts/license/add_license_header.go [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+	if _, err := os.Stat(targetDir); os.IsNotExist(err) {
+		log.Fatalf("Invalid target directory: '%s'\n", targetDir)
+	}
+	fmt.Printf("Processing files in directory: %s\n", targetDir)
+
+	// This map defines the file extensions and their corresponding comment prefixes
+	// that will be added to the license header.
+	extensions := map[string]string{
+		".go":         "//",
+		"Jenkinsfile": "//",
```
[the remark this answers is not in the document]
Comment (facuMH) [reply scripts/license/add_license_header.go]: https://github.com/0xsoniclabs/sonic/pull/335/commits/2390eabca2681c2c5cf5aa57d22f9765aa1de10b

File: scripts/license/add_license_header.go [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+	// the scripts ignores everything inside a build directory
+	for _, pathFragment := range ignoredPaths {
+		if strings.Contains(path, pathFragment) {
+			return true
+		}
+	}
+	return false
+}
+
+func matchExtension(path, ext string) bool {
```
[the remark this answers is not in the document]
Comment (facuMH) [reply scripts/license/add_license_header.go]: https://github.com/0xsoniclabs/sonic/pull/335/commits/579225fb458995e86c307122f399a39414712416

File: scripts/license/add_license_header_test.go [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+				    along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.`
+
+	require.NoError(t, os.WriteFile(originalFileName, []byte(gethHeader), 0644))
+
+	originalContent, err := os.ReadFile(originalFileName)
+	require.NoError(t, err)
+
+	// create a copy of the file with the same content
+	copyName := filepath.Join(tmpDir, "test_geth_license_copy.go")
+	require.NoError(t, os.WriteFile(copyName, originalContent, 0440))
```
[the remark this answers is not in the document]
Comment (facuMH) [reply scripts/license/add_license_header_test.go]: tests have been refactored accordingly
